### PR TITLE
Implement `binomial` and `beta` distribution functions in `keras.random`

### PR DESCRIPTION
**Pull request by KhawajaAbaid**
Following up on the issue https://github.com/keras-team/keras/issues/18918

- Implement `binomial` and `beta` distribution functions in all backends currently supported by Keras namely TensorFlow, Jax, PyTorch and Numpy.
- Add unit tests for each of these functions

Importantly,
As tensorflow doesn't offer a built-in method for beta function so I've implemented a workaround using a statistical formula to use gamma distributed random variables to derive beta distributed random variable.
Specifically, $U(a, b) = X(a) / (X(a) + Y(b))$ where $U(a,b)$ is the beta distributed random variable using parameters $a$ and $b$ and $X(a)$ and $Y(b)$ are gamma-distributed random variables using parameter $a$ and $b$ respectively.

### CODE REVIEW

File: keras/backend/tensorflow/random.py [hunk 1]
```diff
@@ -105,3 +105,44 @@ def gamma(shape, alpha, dtype=None, seed=None):
...
+
+    # Additionally, we'll use two different seeds for our two
+    # gamma random variables to prevent any unintended
+    # dependencies and correlations between the generated values
+    # due to the usage of same seed.
+    seed_1 = tf_draw_seed(seed)
+    # The choice of 12 is totally arbitrary, as we're
+    # incrementing the user specified seed by a CONSTANT to
+    # ensure deterministic results.
+    seed_2 = tf_draw_seed(seed) + 12
```
Comment (fchollet) [inline keras/backend/tensorflow/random.py:137]: Each draw_seed call returns a new seed in the typical case where `seed` is a `SeedGenerator` (which is sequentially deterministic on a single host, but many not be deterministic a in multi host setting). So the + 12 achieves nothing. You could reuse the same seed and add 12 if you only want to make a single call to `draw_seed`.

File: keras/random/random.py [hunk 2]
```diff
@@ -231,3 +231,62 @@ def gamma(shape, alpha, dtype=None, seed=None):
...
+            across multiple calls. To get different random values
+            across multiple calls, use as seed an instance
+            of `keras.random.SeedGenerator`.
+    """
+    return backend.random.binomial(shape, counts=counts, probs=probs,
+                                   dtype=dtype, seed=seed
+                                   )
+
+
+def beta(shape, a, b, dtype=None, seed=None):
```
Comment (fchollet) [inline keras/random/random.py:267]: a -> alpha (it's described as alpha in the docstring and we use `alpha` elsewhere, e.g. for `gamma`
b -> beta

File: keras/random/random.py [hunk 3]
```diff
@@ -231,3 +231,62 @@ def gamma(shape, alpha, dtype=None, seed=None):
...
+    """Draw samples from a Binomial distribution.
+
+    The values are drawn from a binomial distribution with
+    specified count and probability of success parameters.
+
+    Args:
+        shape: The shape of the random values to generate.
+        counts: A number or array of numbers representing the
+            number of trials. It must be broadcastable with probs.
+        probs: A float or array of floats representing the
```
Comment (fchollet) [inline keras/random/random.py:246]: probs -> probabilities (we avoid abbreviations as much as possible, and there's no standard API to follow here since each framework uses a different name)

File: keras/random/random_test.py [hunk 4]
```diff
@@ -257,3 +257,27 @@ def test_gamma(self, seed, shape, alpha, dtype):
...
+        {"seed": 10, "shape": (5, 2), "counts": 1., "probs": 0.5, "dtype": "float16"},
+        {"seed": 10, "shape": (2,), "counts": 1, "probs": 0.5, "dtype": "float32"},
+        {"seed": 10, "shape": (2, 3), "counts": np.array([[1, 2, 3], [4, 5, 6]]),
+         "probs": np.array([[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]]), "dtype": "float32"},
+    )
+    def test_binomial(self, seed, shape, counts, probs, dtype):
+        values = random.binomial(shape, counts=counts,
+                                 probs=probs, seed=seed, dtype=dtype)
+        self.assertEqual(ops.shape(values), shape)
+        self.assertEqual(backend.standardize_dtype(values.dtype), dtype)
```
Comment (fchollet) [inline keras/random/random_test.py:271]: This makes no correctness check. Can we add one (even if statistical in nature)

File: keras/random/random_test.py [hunk 5]
```diff
@@ -257,3 +257,27 @@ def test_gamma(self, seed, shape, alpha, dtype):
...
+        {"seed": 10, "shape": (5, 2), "a": 0.5, "b": 0.9, "dtype": "float16"},
+        {"seed": 10, "shape": (2,), "a": 3., "b": 5., "dtype": "float32"},
+        {"seed": 10, "shape": (2, 3), "a": 1.0, "b": 1.0, "dtype": "float32"},
+    )
+    def test_beta(self, seed, shape, a, b, dtype):
+        values = random.beta(shape, a=a, b=b, seed=seed, dtype=dtype)
+        self.assertEqual(ops.shape(values), shape)
+        self.assertEqual(backend.standardize_dtype(values.dtype), dtype)
+        self.assertGreaterEqual(np.min(ops.convert_to_numpy(values)), b=0.)
+        self.assertLessEqual(np.max(ops.convert_to_numpy(values)), b=1.)
```
Comment (fchollet) [inline keras/random/random_test.py:283]: Same here, this only checks a range. Since we have a manual implementation for TF `beta` it would be good to check correctness.

**Review by fchollet — COMMENTED** [review]
Thank you for the PR!

File: keras/backend/tensorflow/random.py (hunk 1 above)
Comment (KhawajaAbaid) [reply keras/backend/tensorflow/random.py:137]: Thank you so much for the insight. I thought using the same number in `draw_seed` would return the same seed (and even in that case it'd have been better to just call it once, my bad.). Will definitely update it.

File: keras/random/random_test.py (hunk 4 above)
Comment (KhawajaAbaid) [reply keras/random/random_test.py:271]: Yes we definitely can. I'll do that. Sorry at the time I didn't think that testing for the correctness of values generated by an underlying framework would be as important but now that I think about it, it makes loads of sense to do as it'd catch any potential change in the underlying framework or maybe an error in calling the function etc.

**Comment by KhawajaAbaid** [comment]
Again, Thank you so much Mr. Francois Chollet for pointing all my mistakes out and for all the kind suggestions which are helping me learn so much. I've taken careful notice of them and will be updating everything accordingly. Also, I've taken note of the errors, and will resolve those as well.

File: keras/random/random_test.py (hunk 4 above)
Comment (fchollet) [reply keras/random/random_test.py:271]: Maybe we could do a certain number of draws and check the statistical properties of the outputs -- and make the test seeded to avoid flakiness

Comment (KhawajaAbaid) [reply keras/random/random_test.py:271]: Thank you. That makes a lot of sense. Additionally, I'll also do my own research as well to come up with a foolproof way of ensuring the correctness of results for both distributions.

Comment (KhawajaAbaid) [reply keras/random/random_test.py:271]: Hi,
I've come up with two tests to ensure the correctness of binomial values as follows. (I haven't commited/pushed these changes yet)

**Test 1:** Makes sure that the number of times an event occurs doesn't exceed the total possible count for that event specified by the user since the probability value for each event is in $[0, 1]$ interval.
We achieve this by making an elementwise comparison between the `values` array which is generated by the binomial function and contains how many times each event actually occurs with the input argument `counts`  which contains the maximum number of times each event may occur. 

`
        assert np.greater_equal(np.array(counts), np.array(values)).all()
`

Side note: I couldn't find an equivalent implementation for this test case in keras or numpy. So currently I'm using bare `assert` with numpy's method in the `random_test.py ` file.


**Test 2:** Computes the probability of each event using the generated values and compares it with the corresponding input probability value in the `probabilities` input argument to make sure they approximate the input probabilities.
We achieve this by performing an elementwise division of `values` by `counts` and then use the `self.assertAllClose` with `rtol` and `atol` values of `0.005` because if we want to be more precise we'll have to increase the counts value to millions as higher the number of draws, the more closely will the generated probability approximate the input probability of the event. But for the sake of testing, I believe checking to this precision should be fine as we're already performing tens of thousands of draws. 
`
        generated_probabilities = np.array(values) / np.array(counts)
        probabilities = np.ones(shape) * np.array(probabilities)
        self.assertAllClose(
            probabilities, generated_probabilities, rtol=0.005, atol=0.005
        )
`

The current number of draws we're doing in the unit tests:
Unit test 1: `"counts": 5e4`,
Unit test 2: `"counts: 1e5`,
Unit test 3: `"counts": [[1e5, 2e5, 3e5], [4e5, 5e5, 6e5]]`


I've extensively checked these two tests for all four backends with different parameters and they seem to work well. But please let me know if you spot any issues or room for improvement.

File: keras/random/random_test.py (hunk 5 above)
Comment (KhawajaAbaid) [reply keras/random/random_test.py:283]: Hi, 
like you pointed out, merely checking the minimum and maximum value isn't enough since we have a manual implementation for TensorFlow backend so I've come up with the following two tests based on the properties of the beta distribution to ensure that the generated values follow the beta distribution.


**Test 1:** Mean check.
For a beta-distributed random variable, the mean is given by the following formula:
$mean = alpha /  (alpha + beta)$
So the mean of our generated `values` array should ideally be equal to this expected mean. But again, as there's a similar tradeoff we make between the precision and the number of samples we draw, the higher the number of samples, the closer the value of the computed mean will be to the expected mean. So, for the sake of testing, we check the equivalence to two decimal places in the case of a single floating point value, and check the closeness with a `rtol` and `atol` values of `0.005` in the case of an array, which in my opinion should be enough as we run not one but three different unit tests so we're impermeable to flukes.

`
       _alpha_is_an_array = False
        if isinstance(alpha, list):
            alpha = np.array(alpha)
            beta = np.array(beta)
            _alpha_is_an_array = True

        expected_mean = alpha / (alpha + beta)

        if _alpha_is_an_array:
            actual_mean = np.mean(np.array(values), axis=0)
            self.assertAllClose(
                expected_mean.flatten(), actual_mean, atol=0.005, rtol=0.005
            )
        else:
            actual_mean = np.mean(np.array(values).flatten())
            self.assertAlmostEqual(expected_mean, actual_mean, decimal=2)
`


**Test 2:** Variance check.
For a beta-distributed random variable, the variance is given by the following formula:
$variance = (alpha * beta) / ((alpha + beta)^2 * (alpha + beta + 1))$
The implementation logic is similar to the mean above.

`
		
		expected_variance = (alpha * beta) / (
            np.square(alpha + beta) * (alpha + beta + 1)
		)

		if _alpha_is_an_array:
            actual_variance = np.var(np.array(values), axis=0)
            self.assertAllClose(
                expected_variance.flatten(),
                actual_variance,
                atol=0.005,
                rtol=0.005,
            )
        
		else:
            actual_variance = np.var(np.array(values).flatten())
            self.assertAlmostEqual(
                expected_variance, actual_variance, decimal=2
            )
`

I've run these test for all backends and they seem to work well. But if you spot any issues or room for improvement please let me know.
Thanks!

Reference for beta distribution properties: https://stats.libretexts.org/Courses/Saint_Mary's_College_Notre_Dame/MATH_345__-_Probability_(Kuter)/4%3A_Continuous_Random_Variables/4.8%3A_Beta_Distributions#:~:text=distributions%20without%20proof.-,Properties%20of%20Beta%20Distributions,-If

Comment (fchollet) [reply keras/random/random_test.py:283]: Sounds good, you can add a mean and variance check, and make sure to call `set_random_seed()` to ensure that the test will always pass (otherwise it will have some nonzero likelihood to fail randomly).

Comment (KhawajaAbaid) [reply keras/random/random_test.py]: Thank you! I just did. And I've also committed and pushed the changes, fixing other things like lint errors as well.

**Comment by KhawajaAbaid** [comment]
The most tests ran before I pushed my updated `random_test.py` file. For any new test run now, everything should work fine.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!

**Comment by KhawajaAbaid** [comment]
Oh my god. This is a dream come true for me. Always wanted to contribute to this incredible project but never thought i was capable enough for that. Thank you so much, because without you, your awesome book and this keras library I probably would've given up on deep learning right at the start. Even over these two days, I've learned so many new things thanks to your kind suggestions. It's been a true honor! ❤️🫡

**Comment by fchollet** [comment]
Thank you for taking the time to add this feature! It will be valuable to many Keras users.
